### PR TITLE
Special case some BlockDiagLazyTensor matmul ops

### PR DESCRIPTION
**Pull request by Balandat**
This avoids creating MatMulLazyTensor objects in some cases where matmuls are trivial.

### CODE REVIEW

File: gpytorch/lazy/block_diag_lazy_tensor.py [hunk 1]
```diff
@@ -115,6 +115,18 @@ def inv_quad_logdet(self, inv_quad_rhs=None, logdet=False, reduce_inv_quad=True)
             logdet_res = logdet_res.view(*logdet_res.shape).sum(-1)
         return inv_quad_res, logdet_res
 
+    def matmul(self, other):
+        from .diag_lazy_tensor import DiagLazyTensor
```
Comment (dme65) [inline gpytorch/lazy/block_diag_lazy_tensor.py:119]: Why isn't this imported at the top of the file? Does it cause a circular import or something?

**Review by dme65 — APPROVED** [review]
Not super familiar with some of the lazy tensor internals, but the special cases look reasonable to me. Do we have enough test coverage for the special cases introduced here?

File: gpytorch/lazy/block_diag_lazy_tensor.py (hunk 1 above)
Comment (Balandat) [reply gpytorch/lazy/block_diag_lazy_tensor.py:119]: Yes, since `BlockDiagLazyTensor` is imported top-of-file in `diag_lazy_tensor.py` (see below).

**Comment by Balandat** [comment]
> Do we have enough test coverage for the special cases introduced here?

Good question - They ran successfully downstream in BoTorch where this was used, but it probably makes sense to check test coverage in the units again.

**Comment by gpleiss** [comment]
Some unit tests would be helpful here, especially for the LinearOperator refactor.

**Review by gpleiss — CHANGES_REQUESTED** [review]
Looks good, but a few unit tests would be good for the LinearOperator refactor. I can add those if you want.

**Comment by Balandat** [comment]
Yeah I have some local changes to add tests for matmuls with `DiagLazyTensor` and `BlockDiagLazyTensor` but I'll need a bit of time to clean this up. maybe I'll fix the issues there and push them.

**Comment by Balandat** [comment]
Added unit tests. Making sure those pass actually required special casing some logic for lazy tensors in `InterpolatedLazyTensor` - matmuls with diagonal matrices can be done very easily simply by multiplying the interpolation values (someone should double check this though...). There are probably a number of other places where things could be better optimized...

**Comment by Balandat** [comment]
We should probably also more generally think about whether we want the interface to support rmatmuls with lazy tensors as first-class operations (my vote is a strong yes). But we should probably just do this as part of the linear_operator implementation  / refactor.

**Pull request merged**